File: olorin-server/app/schemas/monthly_analysis.py
"""
Monthly Analysis Schemas

Pydantic schemas for monthly sequential analysis data structures.

Feature: monthly-sequential-analysis
"""

from datetime import datetime
from decimal import Decimal
from typing import List, Optional

from pydantic import BaseModel, Field


class DailyAnalysisResult(BaseModel):
    """Result of a single day's analysis within monthly sequential run."""

    date: datetime = Field(description="The date of this analysis window")
    day_of_month: int = Field(description="Day number (1-31)")

    # Entity counts
    entities_discovered: int = Field(
        default=0, description="Number of entities analyzed this day"
    )

    # Confusion matrix metrics
    tp: int = Field(default=0, description="True Positives")
    fp: int = Field(default=0, description="False Positives")
    tn: int = Field(default=0, description="True Negatives")
    fn: int = Field(default=0, description="False Negatives")

    # Financial metrics
    saved_fraud_gmv: Decimal = Field(
        default=Decimal("0"), description="Saved Fraud GMV for this day"
    )
    lost_revenues: Decimal = Field(
        default=Decimal("0"), description="Lost Revenues for this day"
    )
    net_value: Decimal = Field(
        default=Decimal("0"), description="Net Value for this day"
    )

    # References
    investigation_ids: List[str] = Field(
        default_factory=list, description="Investigation IDs created this day"
    )

    # Timing
    started_at: Optional[datetime] = Field(
        default=None, description="When processing started"
    )
    completed_at: Optional[datetime] = Field(
        default=None, description="When processing completed"
    )
    duration_seconds: Optional[float] = Field(
        default=None, description="Processing duration in seconds"
    )

    class Config:
        """Pydantic config."""

        json_encoders = {
            Decimal: lambda v: float(v),
            datetime: lambda v: v.isoformat(),
        }


class MonthlyAnalysisResult(BaseModel):
    """Aggregated result for full month analysis."""

    year: int = Field(description="Year of analysis")
    month: int = Field(description="Month of analysis (1-12)")
    month_name: str = Field(description="Full month name")

    # Status
    total_days: int = Field(description="Total days in month")
    completed_days: int = Field(default=0, description="Days completed")
    is_complete: bool = Field(default=False, description="Whether month is complete")

    # Daily breakdown
    daily_results: List[DailyAnalysisResult] = Field(
        default_factory=list, description="Results for each day"
    )

    # Aggregated confusion matrix
    total_tp: int = Field(default=0, description="Total True Positives")
    total_fp: int = Field(default=0, description="Total False Positives")
    total_tn: int = Field(default=0, description="Total True Negatives")
    total_fn: int = Field(default=0, description="Total False Negatives")

    # Aggregated financial metrics
    total_saved_fraud_gmv: Decimal = Field(
        default=Decimal("0"), description="Total Saved Fraud GMV"
    )
    total_lost_revenues: Decimal = Field(
        default=Decimal("0"), description="Total Lost Revenues"
    )
    total_net_value: Decimal = Field(
        default=Decimal("0"), description="Total Net Value"
    )

    # Derived metrics
    precision: Optional[float] = Field(default=None, description="Precision rate")
    recall: Optional[float] = Field(default=None, description="Recall rate")
    f1_score: Optional[float] = Field(default=None, description="F1 Score")
    roi_percentage: Optional[float] = Field(default=None, description="ROI percentage")

    # Entity counts
    total_entities: int = Field(
        default=0, description="Total entities analyzed across month"
    )

    # Timing
    started_at: Optional[datetime] = Field(
        default=None, description="When monthly analysis started"
    )
    completed_at: Optional[datetime] = Field(
        default=None, description="When monthly analysis completed"
    )

    def calculate_derived_metrics(self) -> None:
        """Calculate precision, recall, F1, and ROI from raw metrics."""
        # Precision: TP / (TP + FP)
        if self.total_tp + self.total_fp > 0:
            self.precision = self.total_tp / (self.total_tp + self.total_fp)

        # Recall: TP / (TP + FN)
        if self.total_tp + self.total_fn > 0:
            self.recall = self.total_tp / (self.total_tp + self.total_fn)

        # F1 Score: 2 * (Precision * Recall) / (Precision + Recall)
        if self.precision and self.recall and (self.precision + self.recall) > 0:
            self.f1_score = (
                2 * (self.precision * self.recall) / (self.precision + self.recall)
            )

        # ROI: (Net Value / Lost Revenues) * 100
        if self.total_lost_revenues > 0:
            self.roi_percentage = float(
                (self.total_net_value / self.total_lost_revenues) * 100
            )

    def aggregate_from_daily(self) -> None:
        """Aggregate metrics from daily results."""
        self.completed_days = len(self.daily_results)
        self.is_complete = self.completed_days >= self.total_days

        self.total_tp = sum(d.tp for d in self.daily_results)
        self.total_fp = sum(d.fp for d in self.daily_results)
        self.total_tn = sum(d.tn for d in self.daily_results)
        self.total_fn = sum(d.fn for d in self.daily_results)

        self.total_saved_fraud_gmv = sum(
            d.saved_fraud_gmv for d in self.daily_results
        )
        self.total_lost_revenues = sum(d.lost_revenues for d in self.daily_results)
        self.total_net_value = sum(d.net_value for d in self.daily_results)

        self.total_entities = sum(d.entities_discovered for d in self.daily_results)

        self.calculate_derived_metrics()

    class Config:
        """Pydantic config."""

        json_encoders = {
            Decimal: lambda v: float(v),
            datetime: lambda v: v.isoformat(),
        }


class MonthlyAnalysisState(BaseModel):
    """Checkpoint state for resumable monthly analysis."""

    year: int = Field(description="Year being analyzed")
    month: int = Field(description="Month being analyzed (1-12)")
    last_completed_day: int = Field(
        default=0, description="Last successfully completed day"
    )

    # Timing
    started_at: datetime = Field(description="When analysis started")
    updated_at: datetime = Field(description="When checkpoint was last updated")

    # Accumulated results
    daily_results: List[DailyAnalysisResult] = Field(
        default_factory=list, description="Results for completed days"
    )

    # Status
    is_complete: bool = Field(default=False, description="Whether analysis finished")
    error_message: Optional[str] = Field(
        default=None, description="Error message if failed"
    )

    class Config:
        """Pydantic config."""

        json_encoders = {
            Decimal: lambda v: float(v),
            datetime: lambda v: v.isoformat(),
        }

File: olorin-fraud/frontend/src/microservices/investigation/types/monthlyAnalysis.ts
/**
 * Monthly Analysis Types
 * Feature: monthly-frontend-trigger
 *
 * TypeScript interfaces for monthly analysis API requests and responses.
 * Matches backend Pydantic schemas in monthly_analysis_api.py.
 *
 * SYSTEM MANDATE Compliance:
 * - No hardcoded values
 * - Complete type definitions
 */

/**
 * Status of a monthly analysis run
 */
export type MonthlyAnalysisRunStatus =
  | 'pending'
  | 'running'
  | 'completed'
  | 'failed'
  | 'cancelled';

/**
 * Request to trigger a new monthly analysis run
 */
export interface MonthlyAnalysisTriggerParams {
  /** Target year for analysis (2020-2030) */
  year: number;
  /** Target month (1-12) */
  month: number;
  /** Day to start/resume from (1-31) */
  resumeFromDay?: number;
  /** Override top percentage of entities to analyze (0.01-1.0) */
  topPercentage?: number;
  /** Override time window in hours (1-168) */
  timeWindowHours?: number;
  /** Run blindspot analysis after monthly flow */
  includeBlindspotAnalysis?: boolean;
}

/**
 * Progress metrics for an in-progress analysis
 */
export interface MonthlyAnalysisProgressMetrics {
  /** Current day being processed */
  currentDay: number;
  /** Total days to process */
  totalDays: number;
  /** Number of days completed */
  daysCompleted: number;
  /** Overall progress percentage (0-100) */
  progressPercentage: number;
  /** Total entities processed so far */
  entitiesProcessed: number;
  /** Total investigations created so far */
  investigationsCreated: number;
}

/**
 * Status response for a monthly analysis run
 */
export interface MonthlyAnalysisStatus {
  /** Unique identifier for this run */
  runId: string;
  /** Current run status */
  status: MonthlyAnalysisRunStatus;
  /** Analysis year */
  year: number;
  /** Analysis month (1-12) */
  month: number;
  /** Full month name (e.g., "January") */
  monthName: string;
  /** When the run started */
  startedAt: string;
  /** When status was last updated */
  updatedAt: string;
  /** When the run completed (if finished) */
  completedAt?: string;
  /** Progress metrics (if running) */
  progress?: MonthlyAnalysisProgressMetrics;
  /** Error message (if failed) */
  errorMessage?: string;
  /** User who triggered the run */
  triggeredBy?: string;
}

/**
 * Summary of a single day's analysis result
 */
export interface DailyResultSummary {
  /** Day of month */
  day: number;
  /** Full date string (YYYY-MM-DD) */
  date: string;
  /** Entities analyzed this day */
  entitiesAnalyzed: number;
  /** Investigations created */
  investigationsCount: number;
  /** True Positives */
  tp: number;
  /** False Positives */
  fp: number;
  /** True Negatives */
  tn: number;
  /** False Negatives */
  fn: number;
  /** Net value for the day */
  netValue: number;
  /** Processing duration in seconds */
  durationSeconds?: number;
}

/**
 * Aggregated metrics for a completed monthly analysis
 */
export interface MonthlyAnalysisMetrics {
  /** Total entities analyzed */
  totalEntities: number;
  /** Total investigations created */
  totalInvestigations: number;
  /** Total True Positives */
  totalTp: number;
  /** Total False Positives */
  totalFp: number;
  /** Total True Negatives */
  totalTn: number;
  /** Total False Negatives */
  totalFn: number;
  /** Precision rate (0-1) */
  precision?: number;
  /** Recall rate (0-1) */
  recall?: number;
  /** F1 Score (0-1) */
  f1Score?: number;
  /** Total saved fraud GMV */
  totalSavedFraudGmv: number;
  /** Total lost revenues */
  totalLostRevenues: number;
  /** Total net value */
  totalNetValue: number;
  /** ROI percentage */
  roiPercentage?: number;
}

/**
 * Full results of a completed monthly analysis run
 */
export interface MonthlyAnalysisResults {
  /** Unique run identifier */
  runId: string;
  /** Run status */
  status: MonthlyAnalysisRunStatus;
  /** Analysis year */
  year: number;
  /** Analysis month */
  month: number;
  /** Full month name */
  monthName: string;
  /** When the run started */
  startedAt: string;
  /** When the run completed */
  completedAt?: string;
  /** Aggregated metrics */
  metrics?: MonthlyAnalysisMetrics;
  /** Results for each day */
  dailyResults: DailyResultSummary[];
}

/**
 * Summary item for history listing
 */
export interface MonthlyAnalysisHistoryItem {
  /** Unique run identifier */
  runId: string;
  /** Analysis year */
  year: number;
  /** Analysis month */
  month: number;
  /** Full month name */
  monthName: string;
  /** Run status */
  status: MonthlyAnalysisRunStatus;
  /** When the run started */
  startedAt: string;
  /** When the run completed */
  completedAt?: string;
  /** Days completed */
  daysCompleted: number;
  /** Total days in month */
  totalDays: number;
  /** Total entities analyzed */
  totalEntities: number;
  /** User who triggered */
  triggeredBy?: string;
}

/**
 * Paginated history response
 */
export interface MonthlyAnalysisHistoryResponse {
  /** List of runs */
  runs: MonthlyAnalysisHistoryItem[];
  /** Total number of runs */
  total: number;
  /** Current page */
  page: number;
  /** Items per page */
  pageSize: number;
  /** More items available */
  hasMore: boolean;
}

/**
 * Link to a generated report
 */
export interface ReportLink {
  /** Type of report (html, csv, pdf) */
  reportType: string;
  /** Download URL */
  url: string;
  /** Suggested filename */
  filename: string;
  /** When report was generated */
  generatedAt: string;
  /** File size in bytes */
  sizeBytes?: number;
}

/**
 * Available reports for a completed run
 */
export interface MonthlyAnalysisReportsResponse {
  /** Run identifier */
  runId: string;
  /** Available reports */
  reports: ReportLink[];
}

/**
 * Request for standalone blindspot analysis
 */
export interface BlindspotAnalysisTriggerParams {
  /** Start date for analysis scope */
  startDate?: string;
  /** End date for analysis scope */
  endDate?: string;
  /** Export results to CSV */
  exportCsv?: boolean;
}

/**
 * Response from blindspot analysis
 */
export interface BlindspotAnalysisResponse {
  /** Analysis status */
  status: string;
  /** Number of blindspots found */
  blindspotsCount: number;
  /** Path to exported CSV */
  csvPath?: string;
  /** Analysis period description */
  analysisPeriod?: string;
}

/**
 * Response after cancelling a run
 */
export interface CancelRunResponse {
  /** Run identifier */
  runId: string;
  /** New status (cancelled) */
  status: MonthlyAnalysisRunStatus;
  /** Cancellation message */
  message: string;
  /** When the run was cancelled */
  cancelledAt: string;
}

/**
 * Terminal statuses (analysis complete)
 */
export const TERMINAL_STATUSES: MonthlyAnalysisRunStatus[] = [
  'completed',
  'failed',
  'cancelled',
];

/**
 * Check if status is terminal (analysis complete)
 */
export function isTerminalStatus(status: MonthlyAnalysisRunStatus): boolean {
  return TERMINAL_STATUSES.includes(status);
}

/**
 * Get display color class for status
 */
export function getStatusColorClass(status: MonthlyAnalysisRunStatus): string {
  switch (status) {
    case 'pending':
      return 'text-yellow-400';
    case 'running':
      return 'text-blue-400';
    case 'completed':
      return 'text-green-400';
    case 'failed':
      return 'text-red-400';
    case 'cancelled':
      return 'text-gray-400';
    default:
      return 'text-gray-400';
  }
}

/**
 * Get display label for status
 */
export function getStatusLabel(status: MonthlyAnalysisRunStatus): string {
  switch (status) {
    case 'pending':
      return 'Pending';
    case 'running':
      return 'Running';
    case 'completed':
      return 'Completed';
    case 'failed':
      return 'Failed';
    case 'cancelled':
      return 'Cancelled';
    default:
      return status;
  }
}
